### PR TITLE
Split bin field from class fields and fix demangling on java

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

This fixes an issue i discovered while refactoring RzBin

We have 2 different types of fields:

1. Binary fields: essentially they are pointer to data structures of the binary itself ([example on elf](https://github.com/rizinorg/rizin/blob/dev/librz/bin/p/bin_elf.inc#L1914))
2. Class fields: as everybody knows you can have fields in classes, but these are mixed with the bin fields.

This PR cleans up these fields and fixes some demangling issues of dex/java

### CODE REVIEW

File: librz/bin/format/dex/dex.c [hunk 1]
```diff
@@ -903,18 +903,34 @@ static ut64 dex_access_flags_to_bin_flags(ut64 access_flags) {
...
+	char *demangled = rz_demangler_java(mangled);
+	free(mangled);
 	return demangled;
 }
 
+static void set_lib_and_class_name(char *mangled, char **out_class, char **out_lib) {
+	if (!mangled) {
+		return;
+	}
+	bool is_java_lang = !strncmp(mangled, "Ljava/lang", strlen("Ljava/lang"));
```
Comment (XVilka) [inline librz/bin/format/dex/dex.c]: Can just use `rz_str_startswith()`

File: librz/bin/format/dex/dex.c [hunk 2]
```diff
@@ -903,18 +903,34 @@ static ut64 dex_access_flags_to_bin_flags(ut64 access_flags) {
...
+
+	char *object = demangle_java_and_free(mangled);
+	if (!object) {
+		return;
+	}
+
+	*out_class = object;
+	if (!is_java_lang || !strncmp(object, "java.lang", strlen("java.lang"))) {
```
Comment (XVilka) [inline librz/bin/format/dex/dex.c]: Same

File: librz/bin/format/dex/dex.h [hunk 3]
```diff
@@ -1,5 +1,5 @@
-// SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
-// SPDX-FileCopyrightText: 2021 deroad <wargio@libero.it>
+// SPDX-FileCopyrightText: 2021-2023 RizinOrg <info@rizin.re>
+// SPDX-FileCopyrightText: 2021-2023 deroad <wargio@libero.it>
```
Comment (XVilka) [inline librz/bin/format/dex/dex.h:2]: I guess no need to update copyright year if you didn't change the file?

File: librz/bin/format/java/class_bin.c [hunk 4]
```diff
@@ -471,7 +472,10 @@ RZ_API RZ_OWN char *rz_bin_java_class_super(RZ_NONNULL RzBinJavaClass *bin) {
 		RZ_LOG_ERROR("java bin: unknown super name at constant pool index %u\n", bin->this_class);
 		return strdup("unknown_super");
 	}
-	return java_class_constant_pool_stringify_at(bin, index);
+	char *tmp = java_class_constant_pool_stringify_at(bin, index);
+	char *class_name = rz_str_newf("L%s;", tmp);
```
Comment (XVilka) [inline librz/bin/format/java/class_bin.c:481]: What if it's NULL?

File: librz/bin/format/java/class_bin.c [hunk 5]
```diff
@@ -1046,6 +1050,37 @@ static char *add_class_name_to_name(char *name, char *classname) {
 	return strdup(name);
 }
 
+static char *demangle_java_and_free(char *mangled) {
+	if (!mangled) {
+		return NULL;
+	}
+	if (!strncmp(mangled, "unknown_", strlen("unknown_"))) {
```
Comment (XVilka) [inline librz/bin/format/java/class_bin.c]: rz_str_startswith()

File: librz/bin/format/java/class_bin.c [hunk 6]
```diff
@@ -1046,6 +1050,37 @@ static char *add_class_name_to_name(char *name, char *classname) {
...
+	char *demangled = rz_demangler_java(mangled);
+	free(mangled);
+	return demangled;
+}
+
+static void set_lib_and_class_name(char *mangled, char **out_class, char **out_lib) {
+	if (!mangled) {
+		return;
+	}
+	bool is_java_lang = !strncmp(mangled, "Ljava/lang", strlen("Ljava/lang"));
```
Comment (XVilka) [inline librz/bin/format/java/class_bin.c]: Same

File: librz/bin/format/java/class_bin.c [hunk 7]
```diff
@@ -1046,6 +1050,37 @@ static char *add_class_name_to_name(char *name, char *classname) {
...
+
+	char *object = demangle_java_and_free(mangled);
+	if (!object) {
+		return;
+	}
+
+	*out_class = object;
+	if (is_java_lang && strncmp(object, "java.lang", strlen("java.lang"))) {
```
Comment (XVilka) [inline librz/bin/format/java/class_bin.c]: Same

File: test/db/cmd/cmd_tc [hunk 8]
```diff
@@ -291,15 +291,15 @@ CMDS=ic*
 EXPECT=<<EOF
 fs classes
 "f class.public.Employee @ 0x1000079a8"
-"f field.Employee.isa @ 0x00000000"
+"f field.Employee.isa @ 0xffffffffffffffff"
```
Comment (XVilka) [inline test/db/cmd/cmd_tc:294]: Is this change intended?

**Review by XVilka — CHANGES_REQUESTED** [review]
The new output looks much better 👍 

Please also fix the bindgen:
```
Mismatched function annotation for rz_bin_class_free at <rizin/librz/include/rz_bin.h:829:13> : RZ_API (was None at <rizin/librz/bin/i/private.h:30:13>
Mismatched function type comment for rz_bin_dex_fields at <rizin/librz/bin/format/dex/dex.h:212:42> : <RzBinField *> (was <RzBinClassField *> at <rizin/librz/bin/format/dex/dex.c:1246:47>)
```

File: librz/bin/format/dex/dex.h (hunk 3 above)
Comment (wargio) [reply librz/bin/format/dex/dex.h:2]: hmm right. let me revert it back

File: test/db/cmd/cmd_tc (hunk 8 above)
Comment (wargio) [reply test/db/cmd/cmd_tc:294]: yes, the previous value was set by `RZ_NEW0` this one is because the default value is `UT64_MAX`

File: librz/bin/format/java/class_bin.c (hunk 5 above)
Comment (wargio) [reply librz/bin/format/java/class_bin.c]: `rz_str_startswith` is not optimized by the compiler, that one is.

**Review by XVilka — CHANGES_REQUESTED** [review]
```
[XX] db/formats/dex dexdump like output of the dex data 
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc dexs bins/dex/Hello.dex
-- stdout
--- expected
+++ actual
@@ --1,7 +-1,7 @@
 DEX file header:
-version             : Android 3.2 (API level 13 and earlier)
+version             : Android M3 release (Nov-Dec 2007)
 checksum            : 1b1f79b5
 signature           : 3c99...7f13
 file_size           : 1000

-- stderr
WARNING: No calling convention defined for this file, analysis may be inaccurate.
```

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
```
Mismatched function annotation for rz_bin_class_free at <rizin/librz/include/rz_bin.h:829:13> : RZ_API (was None at <rizin/librz/bin/bfile.c:514:13>
```

**Review by XVilka — APPROVED** [review]
